### PR TITLE
update APIRequestInfo for APIGroup

### PR DESCRIPTION
**Pull request by deads2k**
Update the `APIRequestInfo` and resolver to handle `APIGroup`.  This is needed before the authorizer can be updated to correctly handle groups.

There were a couple bugs with how the existing experimental group and third party groups were wired in, so I've adjusted those too.

@kubernetes/kube-iam

### CODE REVIEW

**Comment by deads2k** [comment]
integration test added

**Comment by liggitt** [comment]
related to https://github.com/kubernetes/kubernetes/issues/14277

File: pkg/apiserver/handlers.go [hunk 1]
```diff
@@ -427,66 +431,73 @@ type APIRequestInfo struct {
...
 // GetAPIRequestInfo returns the information from the http request.  If error is not nil, APIRequestInfo holds the information as best it is known before the failure
 // Valid Inputs:
 // Storage paths
-// /namespaces
-// /namespaces/{namespace}
-// /namespaces/{namespace}/{resource}
-// /namespaces/{namespace}/{resource}/{resourceName}
-// /{resource}
-// /{resource}/{resourceName}
+// /apis/api-group/{version}/namespaces
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: `{api-group}`

**Comment by deads2k** [comment]
@smarterclayton we need the `APIRequestInfo` changes to support API groups downstream.  My changes to the authorization attributes were necessary to write tests that ensured that `APIRequestInfoResolver` wouldn't be broken again by changes in this area again.

**Comment by deads2k** [comment]
> Should api-group be in eagle braces?

Updated in PR.  I can move the commit its in.

**Comment by smarterclayton** [comment]
A few comments, otherwise looks good

File: pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -78,6 +78,8 @@ type Mux interface {
 type APIGroupVersion struct {
 	Storage map[string]rest.Storage
 
+	// Root is the APIPrefix under which this is being served.  It can also be the APIPrefix/APIGroup that is being served
+	// Since the APIGroup may not contain a '/', you can get the APIGroup by parsing from the last '/'
```
Comment (liggitt) [inline pkg/apiserver/apiserver.go]: that assumes APIPrefix doesn't contain a slash either... is that a valid assumption?

Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > that assumes APIPrefix doesn't contain a slash either... is that a valid assumption?

How does splitting from the last slash (not first slash) prevent the APIPrefix from containing a slash?

Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > How does splitting from the last slash (not first slash) prevent the APIPrefix from containing a slash?

Comment added to handle the caveat.  Different fields are different!!!

File: test/integration/auth_test.go [hunk 3]
```diff
@@ -792,6 +792,95 @@ func newAuthorizerWithContents(t *testing.T, contents string) authorizer.Authori
...
+
+	trackingAuthorizer := &trackingAuthorizer{}
+
+	var m *master.Master
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		m.Handler.ServeHTTP(w, req)
+	}))
+	defer s.Close()
+
+	m = master.New(&master.Config{
```
Comment (smarterclayton) [inline test/integration/auth_test.go]: Isn't there a master helper for this now?  Not sufficient?

Comment (deads2k) [inline test/integration/auth_test.go]: > Isn't there a master helper for this now? Not sufficient?

I wasn't aware of it, this startup boilerplate was a copy/paste of nearby tests.  Pointer to the new pattern?

Comment (smarterclayton) [inline test/integration/auth_test.go]: https://github.com/deads2k/kubernetes/blob/remove-unprefixed-access/test/integration/utils.go#L68 but it doesn't let you override authorizer.

Comment (deads2k) [inline test/integration/auth_test.go]: > https://github.com/deads2k/kubernetes/blob/remove-unprefixed-access/test/integration/utils.go#L68 but it doesn't let you override authorizer.

Ok, then insufficient.

**Comment by smarterclayton** [comment]
This is LGTM

File: test/integration/auth_test.go [hunk 4]
```diff
@@ -792,6 +792,95 @@ func newAuthorizerWithContents(t *testing.T, contents string) authorizer.Authori
...
+func (a *trackingAuthorizer) Authorize(attributes authorizer.Attributes) error {
+	a.requestAttributes = append(a.requestAttributes, attributes)
+	return nil
+}
+
+// TestAuthorizationAttributeDetermination tests that authorization attributes are built correctly
+func TestAuthorizationAttributeDetermination(t *testing.T) {
+	framework.DeleteAllEtcdKeys()
+
+	// This file has alice and bob in it.
```
Comment (smarterclayton) [inline test/integration/auth_test.go]: What is this comment?

File: pkg/apiserver/apiserver.go [hunk 5]
```diff
@@ -112,6 +116,38 @@ const (
...
+
+	return ""
+}
+
+func (g *APIGroupVersion) GetAPIVersion() string {
+	return g.Version
+}
+
+func (g *APIGroupVersion) GetAPIRequestInfoResolver() *APIRequestInfoResolver {
+	apiPrefix := g.GetAPIPrefix()
```
Comment (smarterclayton) [inline pkg/apiserver/apiserver.go]: If these methods aren't used anywhere but in newInstallsr, it would be better to have a method newAPIGroupInfoResolver that takes APIGroupVersion and creates a request info resolver.  Adding a bunch of methods on the type is confusing and people might use them.

File: test/integration/auth_test.go (hunk 4 above)
Comment (deads2k) [inline test/integration/auth_test.go]: > What is this comment?

cruft, removed.

File: pkg/apiserver/apiserver.go (hunk 5 above)
Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > If these methods aren't used anywhere but in newInstallsr, it would be better to have a method newAPIGroupInfoResolver that takes APIGroupVersion and creates a request info resolver. Adding a bunch of methods on the type is confusing and people might use them

This type is borked.  See https://github.com/kubernetes/kubernetes/issues/14433.  These methods are required for anyone to get the actual values they need until we refactor.

File: pkg/auth/authorizer/interfaces.go [hunk 6]
```diff
@@ -41,6 +41,9 @@ type Attributes interface {
 
 	// The kind of object, if a request is for a REST object.
 	GetResource() string
+
+	// The group of the resource, if a request is for a REST object.
+	GetResourceGroup() string
```
Comment (liggitt) [inline pkg/auth/authorizer/interfaces.go]: `APIGroup`?

Comment (deads2k) [inline pkg/auth/authorizer/interfaces.go]: > APIGroup?

Sure. Done.

File: pkg/apiserver/apiserver.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/apiserver/apiserver.go]: Who is calling these methods?  I don't see anyone using this but the request info resolver constructor?

**Comment by liggitt** [comment]
Need to check usage from apiserver/proxy.go, which might be depending on the ability to parse without api/{version}.

**Comment by bgrant0607** [comment]
cc @kubernetes/goog-csi

File: pkg/apiserver/apiserver.go [hunk 7]
```diff
@@ -78,6 +78,10 @@ type Mux interface {
 type APIGroupVersion struct {
 	Storage map[string]rest.Storage
 
+	// Root is the APIPrefix under which this is being served.  It can also be the APIPrefix/APIGroup that is being served
```
Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: > It can also be the APIPrefix/APIGroup that is being served
Unfortunately this will be different from what I implemented in #14156. In #14156, Root just contains the APIPrefix. Group will be part of the Version, in the form of "group/version".

File: pkg/apiserver/apiserver.go [hunk 8]
```diff
@@ -112,6 +116,38 @@ const (
 	MaxTimeoutSecs = 600
 )
 
+func (g *APIGroupVersion) GetAPIPrefix() string {
```
Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: GetAPIPrefix and GetAPIGroup will not be needed if #14156 gets in. And we have utility function to split group/version in pkg/api/util/group_version.go

File: pkg/apiserver/apiserver.go (hunk 7 above)
Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > Unfortunately this will be different from what I implemented in #14156. In #14156, Root just contains the APIPrefix. Group will be part of the Version, in the form of "group/version".

I prefer #14444 that keeps separate fields separate, rather than having a combined "version"

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: I keep them in "group/version" because we expect the user to Specify `APIVersion: "group/version"` in their json/yaml config file. I agree splitting it into separate fields is preferable but that would require a lot more changes than #14444 does. May we merge #14156 first (it's quite big and will be painful to rebase), and then split the fields in another PR?

Comment (smarterclayton) [inline pkg/apiserver/apiserver.go]: APIVersion should remain as GroupVersion - that's a core feature.  It's
just APIGroupVersion (used by the installer) that should split them.

Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > May we merge #14156 first (it's quite big and will be painful to rebase), and then split the fields in another PR?

As long as we can agree to do it and I'd rather not make it a requirement to change every instance in every type at the same time.

Comment (deads2k) [inline pkg/apiserver/apiserver.go]: > May we merge #14156 first (it's quite big and will be painful to rebase), and then split the fields in another PR?

Merging the larger change first is fine.

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: Thank you @deads2k. We use "group/version" in many places in #14156 for fast implementation. We should gradually split them into separate fields where we can.

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: @smarterclayton, yes, I understand. What I tried to say was the fields in APIGroupVersion are populated to many places so it would be difficult to rebase #14156 on this change. I agree we should separate group and version where we can.

**Comment by deads2k** [comment]
@smarterclayton refactored to pass opaque `APIRequestInfoResolver` through `APIGroupVersion` to avoid the need to parse the `Version` field.  It also allowed construction of the `APIRequestInfoResolver` to be common.  I decided against a shared copy, just to avoid the risk of mutation.

**Comment by smarterclayton** [comment]
Mutation is such a problem that it's going to be a problem if it occurs.  I don't think duplicating the copy reduces the implications of a mutable resolver.

**Comment by deads2k** [comment]
> Mutation is such a problem that it's going to be a problem if it occurs. I don't think duplicating the copy reduces the implications of a mutable resolver.

Otherwise?  I'd keep the function because thirdpartyregistration wants a copy to and its easier to recreate the resolver than to hold a ref.

**Comment by deads2k** [comment]
@liggitt can you take over for this one since he's out of town?

**Comment by smarterclayton** [comment]
I'll try to take a look at the next place I have Internet

File: pkg/apiserver/apiserver.go [hunk 9]
```diff
@@ -81,6 +80,9 @@ type APIGroupVersion struct {
 	Root    string
 	Version string
 
+	// APIRequestInfoResolver is used to parse URLs for the legacy proxy handler.  Don't use this for anything else
```
Comment (smarterclayton) [inline pkg/apiserver/apiserver.go]: Add TODO: refactor proxy handler to use sub resources

**Comment by smarterclayton** [comment]
Minor nit

**Comment by deads2k** [comment]
@smarterclayton TODO added (next line) and green.
